crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,234 @@
+        let cache: LruCache<[u8; 32], GTElement> =
+            LruCache::new(NonZeroUsize::new(cache_size.unwrap_or(50000)).unwrap());
+        Self { cache }
+    }
+
+    pub fn aggregate_verify<M: AsRef<[u8]>>(
+        &mut self,
+        pks: &[PublicKey],
+        msgs: &[M],
+        sig: &Signature,
+    ) -> bool {
+        let iter = pks.iter().zip(msgs.iter()).map(|(pk, msg)| -> GTElement {
+            let mut hasher = Sha256::new();
+            hasher.update(pk.to_bytes());
+            hasher.update(msg); // pk + msg
+            let h: [u8; 32] = hasher.finalize().into();
+
+            let pairing: Option<GTElement> = self.cache.get(&h).cloned();
+
+            if let Some(pairing) = pairing {
+                // equivalent to `if pairing is not None`
+                return pairing;
+            }
+            // if pairing is None then make pairing and add to cache
+            let mut aug_msg = pk.to_bytes().to_vec();
+            aug_msg.extend_from_slice(msg.as_ref()); // pk + msg
+            let aug_hash: Signature = hash_to_g2(&aug_msg);
+
+            let pairing: GTElement = aug_hash.pair(pk);
+            let mut hasher = Sha256::new();
+            hasher.update(&aug_msg);
+            let h: [u8; 32] = hasher.finalize().into();
+            self.cache.put(h, pairing.clone());
+            pairing
+        });
+        agg_ver_gt(sig, iter)
+    }
+}
+
+// Python Functions
+
+// Commented out for now as we may remove these
+// as the python consensus code that uses it is being ported to rust.
+
+// #[cfg(feature = "py-bindings")]
+// #[pymethods]
+// impl BLSCache {
+//     #[new]
+//     pub fn init() -> Self {
+//         Self::default()
+//     }
+
+//     #[staticmethod]
+//     #[pyo3(name = "generator")]
+//     pub fn py_generator(size: Option<&PyInt>) -> Self {
+//         size.map(|s| Self::new(s.extract().unwrap()))
+//             .unwrap_or_default()
+//     }
+
+//     #[pyo3(name = "aggregate_verify")]
+//     pub fn py_aggregate_verify(
+//         &mut self,
+//         pks: &PyList,
+//         msgs: &PyList,
+//         sig: &Signature,
+//         force_cache: &PyBool,
+//     ) -> PyResult<bool> {
+//         let pks_r: Vec<[u8; 48]> = pks
+//             .iter()
+//             .map(|item| item.extract::<[u8; 48]>())
+//             .collect::<PyResult<_>>()?;
+//         let msgs_r: Vec<&[u8]> = msgs
+//             .iter()
+//             .map(|item| item.extract::<&[u8]>())
+//             .collect::<PyResult<_>>()?;
+//         let force_cache_bool = force_cache.extract::<bool>()?;
+//         Ok(self.aggregate_verify(&pks_r, &msgs_r, sig, force_cache_bool))
+//     }
+
+//     #[pyo3(name = "len")]
+//     pub fn py_len(&self) -> PyResult<usize> {
+//         Ok(self.cache.len())
+//     }
+// }
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use crate::aggregate;
+    use crate::sign;
+    use crate::SecretKey;
+
+    #[test]
+    pub fn test_instantiation() {
```

Comment:
I don't think these tests cover all interesting cases. It would probably easier to read these if the setup-code was factored out, to make each test clearly show the actual function that's being tested.

I would like to see tests where we have 1000 public keys + messages validated with 0% cache hit rate, 100% cache hit rate, and something in between, ensuring we get the same result.

Also some negative tests, ensuring we *catch* invalid signatures (see the tests in `chia-bls`).

When you removed the `force_cache` option, didn't any tests become redundant? Or did we not have tests for that feature before?

Reply:
All test functions should be private. i.e. not `pub`, unless you actually use them from some other file. I don't expect you to though, as that would be a bit odd.